### PR TITLE
tensorflow_c 1.0.0 (new formula)

### PR DESCRIPTION
**Pull request by kali**
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/master/CONTRIBUTING.md)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `brew install <formula>`)?

-----

### CODE REVIEW

File: Formula/tensorflow_c.rb [hunk 1]
```diff
@@ -0,0 +1,42 @@
+class TensorflowC < Formula
+  desc "C interface for Google's OS library for Machine Intelligence"
+  homepage "https://www.tensorflow.org/"
+  url "https://github.com/tensorflow/tensorflow", :using => :git,
```
Comment (bfontaine) [inline Formula/tensorflow_c.rb]: Please add `.git` at the end of the URL and remove `:using => :git` (it’s already the default).

File: Formula/tensorflow_c.rb [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+    :tag => "v1.0.0", :revision => "07bb8ea2379bd459832b23951fb20ec47f3fdbd4"
+
+  depends_on "bazel" => :build
+  depends_on "pkg-config" => :run
+
+  def install
+    system 'echo "\n\n\n\n\n\n\n\n\n" | ./configure'
```
Comment (bfontaine) [inline Formula/tensorflow_c.rb]: Why is this needed?

File: Formula/tensorflow_c.rb [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+    system "bazel", "build", "--compilation_mode=opt", "--copt=-march=native", "tensorflow:libtensorflow_c.so"
+    lib.install "bazel-bin/tensorflow/libtensorflow_c.so"
+    cp "tensorflow/c/c_api.h", "tensorflow_c_api.h"
+    include.install "tensorflow_c_api.h"
```
Comment (bfontaine) [inline Formula/tensorflow_c.rb]: Use `include.install "tensorflow/c/c_api.h" => "tensorflow_c_api.h"` instead.

File: Formula/tensorflow_c.rb [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+    pc = <<-EOF.gsub(/^\s+/, "")
+      Name: tensorflow_c
+      Description: Tensorflow c lib
+      Version: #{version}
+      Libs: -L#{lib} -ltensorflow_c
+      Cflags: -I#{include}
+    EOF
+    mkdir_p(lib/"pkgconfig")
+    File.open(lib/"pkgconfig/tensorflow_c.pc", "w") { |f| f.write(pc) }
```
Comment (bfontaine) [inline Formula/tensorflow_c.rb]: Use `(lib/"pkgconfig/tensorflow_c.pc").write pc` instead. You don’t even need to make a variable for it.

File: Formula/tensorflow_c.rb [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+  depends_on "bazel" => :build
+  depends_on "pkg-config" => :run
+
+  def install
+    system 'echo "\n\n\n\n\n\n\n\n\n" | ./configure'
+    system "bazel", "build", "--compilation_mode=opt", "--copt=-march=native", "tensorflow:libtensorflow_c.so"
+    lib.install "bazel-bin/tensorflow/libtensorflow_c.so"
+    cp "tensorflow/c/c_api.h", "tensorflow_c_api.h"
+    include.install "tensorflow_c_api.h"
+    pc = <<-EOF.gsub(/^\s+/, "")
```
Comment (bfontaine) [inline Formula/tensorflow_c.rb]: Use `EOS.undent` instead (and `EOS` instead of `EOF` at the end).

File: Formula/tensorflow_c.rb [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+      Name: tensorflow_c
+      Description: Tensorflow c lib
+      Version: #{version}
+      Libs: -L#{lib} -ltensorflow_c
+      Cflags: -I#{include}
+    EOF
+    mkdir_p(lib/"pkgconfig")
```
Comment (bfontaine) [inline Formula/tensorflow_c.rb]: You don’t need this.

File: Formula/tensorflow_c.rb [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+    File.open(lib/"pkgconfig/tensorflow_c.pc", "w") { |f| f.write(pc) }
+  end
+
+  test do
+    # test a call on TF_Version(), checking .h, .so, pkg-config setup.
+    File.open("test.c", "w") do |f|
```
Comment (bfontaine) [inline Formula/tensorflow_c.rb]: Use `(testpath/"test.c").write <<-EOS.undent`.

File: Formula/tensorflow_c.rb [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+      #include <stdio.h>
+      #include <tensorflow_c_api.h>
+      int main() {
+        printf("%s\\n", TF_Version());
+      }
+      EOF
+    end
+    system "sh", "-c", "gcc `pkg-config --libs --cflags tensorflow_c` -o test_tf test.c"
+    found_version = `./test_tf`.strip
+    assert found_version == version
```
Comment (bfontaine) [inline Formula/tensorflow_c.rb]: Use `assert_equal version, shell_output("test_tf")`. You don’t need the `.strip` if you remove `\\n` in the C code.

File: Formula/tensorflow_c.rb [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+    File.open("test.c", "w") do |f|
+      f.write(<<-EOF.gsub(/^\s+/, ""))
+      #include <stdio.h>
+      #include <tensorflow_c_api.h>
+      int main() {
+        printf("%s\\n", TF_Version());
+      }
+      EOF
+    end
+    system "sh", "-c", "gcc `pkg-config --libs --cflags tensorflow_c` -o test_tf test.c"
```
Comment (bfontaine) [inline Formula/tensorflow_c.rb]: Use the formula paths instead; formulae should be testable without being linked.

**Review by bfontaine — CHANGES_REQUESTED** [review]

File: Formula/tensorflow_c.rb (hunk 9 above)
Comment (kali) [inline Formula/tensorflow_c.rb]: I wanted to check the generated .pc was fine. If I understand correctly, you'd prefer me to replace the pkg-config call by something like -L#{lib} and so on ?

File: Formula/tensorflow_c.rb (hunk 2 above)
Comment (kali) [inline Formula/tensorflow_c.rb]: This one is very sad. Tensorflow `./configure` is not an autotools like thing. It will expect you to hit return 9 times to accept to default to 9 questions. There is not provision for non interactive run.

Alternatively, I could do a `yes ''| ./configure` but `yes` would fail systematically with a borken pipe when configure ends and it would get tricky to detect a ./configure fail.

If you have a better idea, I'm all hears.

Comment (MikeMcQuaid) [inline Formula/tensorflow_c.rb]: @kali I'd advise submitting a patch upstream (or at least an issue) that fixes this problem. This is particularly hostile to any package manager packaging this software as-is.

File: Formula/tensorflow_c.rb (hunk 9 above)
Comment (bfontaine) [inline Formula/tensorflow_c.rb]: Yes, thanks.

File: Formula/tensorflow_c.rb (hunk 2 above)
Comment (kali) [inline Formula/tensorflow_c.rb]: ok, it is possible to pass environment variables matching the default to get the configure script running with no input. Still very fragile, though.

**Comment by fxcoudert** [comment]
Test is failing with:

```
==> brew test tensorflow_c --verbose
==> FAILED
Testing tensorflow_c
==> Using the sandbox
/usr/bin/sandbox-exec -f /tmp/homebrew20170224-99173-14yz7vt.sb /System/Library/Frameworks/Ruby.framework/Versions/2.0/usr/bin/ruby -W0 -I /usr/local/Homebrew/Library/Homebrew -- /usr/local/Homebrew/Library/Homebrew/test.rb /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/tensorflow_c.rb --verbose
Error: tensorflow_c: failed
Calling ENV.gcc is deprecated!
Use ENV.gcc_4_2 instead.
/usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/tensorflow_c.rb:48:in `block in <class:TensorflowC>'
```

It should probably be `ENV.cc` instead of `ENV.gcc`.

File: Formula/tensorflow_c.rb [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+    ENV["TF_NEED_JEMALLOC"]="1"
+    ENV["TF_NEED_GCP"]="0"
+    ENV["TF_NEED_HDFS"]="0"
+    ENV["TF_ENABLE_XLA"]="0"
+    ENV["USE_DEFAULT_PYTHON_LIB_PATH"]="1"
+    ENV["TF_NEED_OPENCL"]="0"
+    ENV["TF_NEED_CUDA"]="0"
+    system "./configure"
+
+    system "bazel", "build", "--compilation_mode=opt", "--copt=-march=native", "tensorflow:libtensorflow_c.so"
```
Comment (asimshankar) [inline Formula/tensorflow_c.rb]: Could you use `tensorflow:libtensorflow.so` instead?
That is the target that is used as part of the binary releases of the TensorFlow C library.

File: Formula/tensorflow_c.rb [hunk 11]
```diff
@@ -0,0 +1,46 @@
+class TensorflowC < Formula
+  desc "C interface for Google's OS library for Machine Intelligence"
+  homepage "https://www.tensorflow.org/"
+  url "https://github.com/tensorflow/tensorflow.git", :tag => "v1.0.0",
+    :revision => "07bb8ea2379bd459832b23951fb20ec47f3fdbd4"
+
+  depends_on "bazel" => :build
+  depends_on "pkg-config" => :run
+
+  def install
```
Comment (asimshankar) [inline Formula/tensorflow_c.rb]: Can the release tarballs of the TensorFlow C library be used to avoid building from source?

That is recommended for building language bindings for TensorFlow (for example, the [Go bindings](https://github.com/tensorflow/tensorflow/tree/master/tensorflow/go) and hopefully the Rust bindings will switch to using this binary target when possible too)

File: Formula/tensorflow_c.rb [hunk 12]
```diff
@@ -0,0 +1,46 @@
...
+    ENV["PYTHON_BIN_PATH"]=`which python`.strip
+    ENV["CC_OPT_FLAGS"]="-march=native"
+    ENV["TF_NEED_JEMALLOC"]="1"
+    ENV["TF_NEED_GCP"]="0"
+    ENV["TF_NEED_HDFS"]="0"
+    ENV["TF_ENABLE_XLA"]="0"
+    ENV["USE_DEFAULT_PYTHON_LIB_PATH"]="1"
+    ENV["TF_NEED_OPENCL"]="0"
+    ENV["TF_NEED_CUDA"]="0"
```
Comment (asimshankar) [inline Formula/tensorflow_c.rb]: Would it make sense for these to be options, so that GPU support etc. can be enabled with flags like `--with-cuda` or something to the `brew install` command?

**Review by asimshankar — COMMENTED** [review]
A TensorFlow team member chiming in here.

Firstly, this is great! Thanks for doing this, adding a brew formula for the TensorFlow C library has been on our TODO list for a while. I was hoping to get to it over the next couple of weeks.

A few comments below. Also, would `libtensorflow` be a more appropriate name for this formula?

Thanks much,

-- Asim

File: Formula/tensorflow_c.rb (hunk 11 above)
Comment (MikeMcQuaid) [inline Formula/tensorflow_c.rb]: No, sorry. Homebrew requires building from source.

File: Formula/tensorflow_c.rb (hunk 12 above)
Comment (MikeMcQuaid) [inline Formula/tensorflow_c.rb]: We'd rather minimise our number of options until requested by users.

File: Formula/tensorflow_c.rb (hunk 10 above)
Comment (MikeMcQuaid) [inline Formula/tensorflow_c.rb]: It seems like `libtensorflow.dylib` should be used as this is macOS where `.so` indicates a plugin rather than library.

Comment (asimshankar) [inline Formula/tensorflow_c.rb]: Oh, my comment was about the bazel targets and not the filename.
I was suggesting that the build command use the bazel target [`//tensorflow:libtensorflow.so`](https://github.com/tensorflow/tensorflow/blob/e75c0fc/tensorflow/BUILD#L300) instead of [`//tensorflow:libtensorflow_c.so`](https://github.com/tensorflow/tensorflow/blob/e75c0fc/tensorflow/BUILD#L309) (which is what is being done right now)

File: Formula/tensorflow_c.rb [hunk 13]
```diff
@@ -0,0 +1,46 @@
...
+    system "./configure"
+
+    system "bazel", "build", "--compilation_mode=opt", "--copt=-march=native", "tensorflow:libtensorflow_c.so"
+    lib.install "bazel-bin/tensorflow/libtensorflow_c.so"
+    include.install "tensorflow/c/c_api.h" => "tensorflow_c_api.h"
```
Comment (asimshankar) [inline Formula/tensorflow_c.rb]: Is the renaming necessary?
Can the include be installed in `<include_dir>/tensorflow/c/c_api.h` to avoid the rename?

My concern with renaming the file is that unless the same rename is followed in all other forms of distribution (such as on other platforms/OSes), this rename makes client code less portable.

Comment (MikeMcQuaid) [inline Formula/tensorflow_c.rb]: Agreed.

File: Formula/tensorflow_c.rb (hunk 10 above)
Comment (MikeMcQuaid) [inline Formula/tensorflow_c.rb]: > Oh, my comment was about the bazel targets and not the filename.

Gotcha 👍

**Comment by kali** [comment]
@asimshankar Thanks for your interest, just wanted to make sure you have also seen the rest of our efforts/struggle over there: https://github.com/snipsco/tensorflow-build and there https://medium.com/snips-ai/how-we-made-tensorflow-run-on-a-raspberry-pi-using-rust-7478f7a31329#.kgvgxpech . Feedback welcome :)

File: Formula/tensorflow_c.rb (hunk 10 above)
Comment (kali) [inline Formula/tensorflow_c.rb]: @asimshankar I'm not sure what you recommend here, actually. Are these two libraries the same thing ? I know that the Rust bindings, for instance, looks for `tensorflow_c` in pkg-config. Unless I'm confused, building `//tensorflow:libtensorflow.so` with bazel does not produce `libtensorflow_c.so` .

Comment (asimshankar) [inline Formula/tensorflow_c.rb]: @kali: Yes, these two libraries are the same thing and the fact that there are two blaze targets that do exactly the same thing was an accident. So the idea is to suggest usage of `llibtensorflow.so` and remove `libtensorflow_c.so`.

I recently updated the Rust bindings to use `libtensorflow.so` instead of `libtensorflow_c.so` in https://github.com/tensorflow/rust/pull/56 and https://github.com/tensorflow/rust/pull/60,  and then plan to remove the `libtensorflow_c.so` target completely.

File: Formula/libtensorflow.rb [hunk 14]
```diff
@@ -0,0 +1,46 @@
...
+  def install
+    ENV["PYTHON_BIN_PATH"]=`which python`.strip
+    ENV["CC_OPT_FLAGS"]="-march=native"
+    ENV["TF_NEED_JEMALLOC"]="1"
+    ENV["TF_NEED_GCP"]="0"
+    ENV["TF_NEED_HDFS"]="0"
+    ENV["TF_ENABLE_XLA"]="0"
+    ENV["USE_DEFAULT_PYTHON_LIB_PATH"]="1"
+    ENV["TF_NEED_OPENCL"]="0"
+    ENV["TF_NEED_CUDA"]="0"
```
Comment (bfontaine) [inline Formula/libtensorflow.rb]: Please leave a space before and after `=`. You can also probably use `which("python").to_s` instead of shelling out with `` `which python`.strip ``.

File: Formula/libtensorflow.rb [hunk 15]
```diff
@@ -0,0 +1,46 @@
...
+      Name: tensorflow
+      Description: Tensorflow library
+      Version: #{version}
+      Libs: -L#{lib} -ltensorflow
+      Cflags: -I#{include}
+    EOS
+  end
+
+  test do
+    # test a call on TF_Version()
```
Comment (bfontaine) [inline Formula/libtensorflow.rb]: You can remove this comment.

File: Formula/libtensorflow.rb [hunk 16]
```diff
@@ -0,0 +1,46 @@
+class Libtensorflow < Formula
+  desc "C interface for Google's OS library for Machine Intelligence"
+  homepage "https://www.tensorflow.org/"
+  url "https://github.com/tensorflow/tensorflow.git", :tag => "v1.0.0",
+    :revision => "07bb8ea2379bd459832b23951fb20ec47f3fdbd4"
+
+  depends_on "bazel" => :build
+  depends_on "pkg-config" => :run
```
Comment (bfontaine) [inline Formula/libtensorflow.rb]: Is it really needed?

File: Formula/libtensorflow.rb [hunk 17]
```diff
@@ -0,0 +1,46 @@
+class Libtensorflow < Formula
+  desc "C interface for Google's OS library for Machine Intelligence"
+  homepage "https://www.tensorflow.org/"
+  url "https://github.com/tensorflow/tensorflow.git", :tag => "v1.0.0",
+    :revision => "07bb8ea2379bd459832b23951fb20ec47f3fdbd4"
```
Comment (bfontaine) [inline Formula/libtensorflow.rb]: Can’t we build from the tarball?

**Review by bfontaine — CHANGES_REQUESTED** [review]

File: Formula/libtensorflow.rb (hunk 17 above)
Comment (kali) [inline Formula/libtensorflow.rb]: I could not find a tgz. The "install from source" page pulls from github. @asimshankar, is there something I'm missing ?

Comment (MikeMcQuaid) [inline Formula/libtensorflow.rb]: https://github.com/tensorflow/tensorflow/archive/v1.0.0.tar.gz from https://github.com/tensorflow/tensorflow/releases/tag/v1.0.0

Comment (kali) [inline Formula/libtensorflow.rb]: thx.

**Review by bfontaine — APPROVED** [review]

**Comment by kali** [comment]
Hey there... could we move this forward and merge it ? Is there something more I can / should do ?

**Comment by fxcoudert** [comment]
Thanks @kali for your contribution!

**Comment by carlthome** [comment]
Newbie brew user here!

1. I thought `brew install tensorflow` would get the TensorFlow Python API, not the C API. IMO it would be good to require `brew install libtensorflow` as that's typically what the non-Python stuff is called in TensorFlow's docs.
2. `brew install tensorflow` "successfully" installed the formula despite my system not having Bazel installed.
```sh
macbook:~ carl$ brew install tensorflow
Warning: libtensorflow 1.6.0 is already installed
macbook:~ carl$ bazel
-bash: bazel: command not found
macbook:~ carl$ brew install bazel
bazel: Java 1.8 is required to install this formula.
JavaRequirement unsatisfied!
You can install with Homebrew-Cask:
 brew cask install caskroom/versions/java8
You can download from:
 https://www.oracle.com/technetwork/java/javase/downloads/index.html
Error: An unsatisfied requirement failed this build.
```
Shouldn't the install halt on https://github.com/Homebrew/homebrew-core/blob/master/Formula/libtensorflow.rb#L37 or am I missing something obvious?

**Comment by bfontaine** [comment]
@carlthome you should open a separate issue.
